wagtail/core/collectors.py

```diff
@@ -555,24 +567,35 @@ def from_nested_list(cls, nested_list, parent=None, on_delete=CASCADE,
 
         # Retain the Use record for the most-recently-handled instance, as this will be used as the parent
         # for any sub-lists we encounter
-        main_use = None
+        parent_use = None
 
         for obj in nested_list:
             if isinstance(obj, list):
                 yield from cls.from_nested_list(
-                    obj, parent=main_use, on_delete=on_delete, exclude=exclude)
+                    obj, parent=parent_use, on_delete=on_delete, exclude=exclude)
             else:
                 use = cls(obj, parent=parent, on_delete=on_delete)
+                # When ``use`` was already listed, we don’t want to list it
+                # but we want to mark it as parent of the next item, in case
+                # that next item is a list.
                 if use in exclude:
-                    if use not in originals:
-                        for other_use in exclude:
-                            if use == other_use:
-                                main_use = other_use
-                                break
+                    # When ``use`` is in ``originals``, we don’t want
+                    # to mark it as parent of the next item because we want
+                    # to display the next item as root, otherwise the first
+                    # level of deleted objects would be shown nested.
```

Comment:
@BertrandBordage OK, let me see if I've got this right: this logic comes into play when `nested_list` is a structure such as:

    [a, [b], b, [c]]

where:

* `b` is one of the items from `originals`
* `b` appears in the child list of an earlier item
* the top-level occurrence of `b` in the list is immediately followed by a sub-list

In this case, we do _not_ set `parent_use` to `b` - it gets left at its existing value, which is `a`, and so `c` ends up being assigned as a child of `a`. Is that what you intended to happen? Is this _actually_ a legitimate way that NestedObjects might represent `c` being a child of `a`, and are we relying on the fact that if `c` was meant to be a child of `b`, the list would look like `[a, [b, [c]], b, [c]]` (in which case the second `c` would be skipped along with the second `b`)? Ouch...

Reply:
@BertrandBordage Ah, I missed the detail that `exclude` always contains the originals whenever we call this - lines 565-566 (which are never used?) led me to understand that the initial state was the empty set. Could we make `exclude` a required argument, or at least change those lines to

    if exclude is None:
        exclude = set(originals)

so that it's not misleading?

Reply:
We can indeed add it. I kept it consistent with `from_flat_iterable`, but that was probably not the best decision.